File: src/dailyGroupWars/June22/424.cc
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
  int characterReplacement(string s, int k) {
    int l = 0;
    int maxLength = 1;
    int currentLength = 1;
    char prev = s[0];
    int mostFreq = 0;
    std::unordered_map<int, int> fmap;
    for (int r = 0; r < s.size(); r++) {
      fmap[s[r]]++;
      int window = r - l + 1;
      for (auto &kv : fmap) {
        mostFreq = max(mostFreq, kv.second);
      }
      if ((window - mostFreq) <= k) {
        maxLength = max(maxLength, window);
        // re-compute the most frequent
      } else {
        fmap[s[l]]--;
        l++;
      }
    }
    return maxLength;
  }
};

File: src/archive/Zon/AmazonPrepOld/SlidingWindow/424.go
package SlidingWindow

// time complexity -> O(n)
func characterReplacement(s string, k int) int {
    maxLength := 0
    l := 0
    maxFreq := 0
    freqMap := make(map[rune]int)

    for r := 0; r < len(s); r++ {
        // re-compute most frequent character in current window
        freqMap[rune(s[r])]++
        maxFreq = max(maxFreq, freqMap[rune(s[r])])
        
        for r - l + 1 - maxFreq > k {
            freqMap[rune(s[l])]--
            l++
        }
        
        maxLength = max(maxLength, r - l + 1)
    }
    return maxLength
}
